### PR TITLE
feature(zy): add count-based pixelcnn

### PR DESCRIPTION
**Pull request by youngzhou1999**

### CODE REVIEW

File: ding/reward_model/count_based_model.py [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+from .base_reward_model import BaseRewardModel
+
+
+@REWARD_MODEL_REGISTRY.register('countbased')
+class CountbasedRewardModel(BaseRewardModel):
+    """
+        # TODO:
+    """
+    
+    config = dict(
```
Comment (PaParaZz1) [inline ding/reward_model/count_based_model.py:21]: There should be some comments in default config, like that in policy definition

File: ding/reward_model/count_based_model.py [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+            self._counter.to(self.device)
+            self.intrinsic_reward_type = cfg.intrinsic_reward_type
+            self.obs_shape = (cfg.img_height, cfg.img_width)
+            self.index_range = np.arange(cfg.img_height * cfg.img_width)
+            assert self.intrinsic_reward_type in ['add', 'new', 'assign']
+            self.train_data = []
+            self.opt = optim.RMSprop(
+                self._counter.parameters(),
+                momentum=0.9,
+                weight_decay=0.95,
```
Comment (PaParaZz1) [inline ding/reward_model/count_based_model.py:74]: Why use 0.95 weight_decay here, usually it is a small positive value (e.g.: 1e-3 or 1e-4)

File: ding/reward_model/count_based_model.py [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+                eps=1e-4,
+            )
+
+    def _train(self):
+        train_data: list = self.train_data
+        train_data: torch.Tensor = torch.stack(train_data).to(self.device)
+        flattened_logits, target_pixel_loss, _ = self._counter(train_data)
+        loss = nn.CrossEntropyLoss()(flattened_logits, target_pixel_loss.long())
+        loss = loss.mean()
```
Comment (PaParaZz1) [inline ding/reward_model/count_based_model.py:83]: There is no need to call `mean` method here, `nn.CrossEntropyLoss()` returns 0-dim tensor by default

File: ding/reward_model/count_based_model.py [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+        pred_prob = target[self.index_range, indexes.long()]
+        return pred_prob
+
+    def _collect_states(self, data):
+        obs = []
+        for item in data:
+            state = item['obs']
+            state = state.permute(1, 2, 0)
+            state = state.numpy()
+            state = cv2.resize(state, self.obs_shape)
```
Comment (PaParaZz1) [inline ding/reward_model/count_based_model.py:130]: Use `torch.nn.functional.interpolate` to replace `cv2.resize`, so you don't need transform to numpy

File: ding/reward_model/count_based_model.py [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+        self.train_data.clear()
+
+            
+
+
+
+# Borrowed implement
+
+class GatedActivation(nn.Module):
+    """Gated activation function as introduced in [2].
```
Comment (PaParaZz1) [inline ding/reward_model/count_based_model.py:152]: Add reference for `[2]`

File: ding/reward_model/count_based_model.py [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+    The function computes actiation_fn(f) * sigmoid(g). The f and g correspond to the
+    top 1/2 and bottom 1/2 of the input channels.
+    """
+
+    def __init__(self, activation_fn=torch.tanh):
```
Comment (PaParaZz1) [inline ding/reward_model/count_based_model.py:157]: Use `torch.nn.Tanh()` to replace `torch.tanh`, so we can add `print(model)` to know what is applied in this activation. For the details, please refer to `__repr__` method of `nn.Module`

File: ding/reward_model/count_based_model.py [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+        hstack = self._hstack_residual(hstack)
+        # NOTE(eugenhotaj): We cannot use a residual connection for causal layers
+        # otherwise we'll have access to future pixels.
+        if not self._mask_center:
+            hstack += hstack_input
+
+        return vstack, hstack, skip
+
+
+class GatedPixelCNN(nn.Module):
```
Comment (PaParaZz1) [inline ding/reward_model/count_based_model.py:273]: Try to write unittest for `GatedPixelCNN`, you can refer to unittest example [here](https://github.com/opendilab/DI-engine/blob/main/ding/model/common/tests/test_encoder.py#L13)

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/reward_model/count_based_model.py (hunk 2 above)
Comment (youngzhou1999) [reply ding/reward_model/count_based_model.py:74]: the paper uses decay=0.95,  it can be seen in page 11 of https://arxiv.org/pdf/1703.01310.pdf.

File: ding/reward_model/count_based_model.py (hunk 3 above)
Comment (youngzhou1999) [reply ding/reward_model/count_based_model.py:83]: In this implement, the loss is [42 x 42] for every pixel in the input, so I think it needs mean.

File: ding/reward_model/count_based_model.py (hunk 6 above)
Comment (youngzhou1999) [reply ding/reward_model/count_based_model.py:157]: got it, thanks.

File: ding/reward_model/count_based_model.py [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+
+            prob = (self._probs(obs) + 1e-8).sum().item()
+
+            with torch.no_grad():
+                recoding_prob = (self._probs(obs) + 1e-8).sum().item()
+
+            pred_gain = np.log(recoding_prob) - np.log(prob)
+
+            intrinsic_reward = [pow(
+                exp(self.cfg.bonus_coeffient * pow(t+1, -0.5) * pred_gain) - 1, 0.5
```
Comment (puyuan1996) [inline ding/reward_model/count_based_model.py:108]: why t+1 not t, would you please give some annotations here?

File: ding/reward_model/count_based_model.py [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+            vstack, hstack, skip = gated_layer(vstack, hstack)
+            skip_connections += skip
+            # print(f'3x3 gate:{skip_connections.shape}')
+        # return self._head(skip_connections) # logits
+        logits = self._head(skip_connections)
+        logits, x = logits.permute(0, 2, 3, 1), x.permute(0, 2, 3, 1)
+        # print(f'logits, x:{logits.shape, x.shape}')
+        flattened_logits = torch.reshape(logits, [-1, self.q_level])
+        # print(f'flat logits:{flattened_logits.shape}')
+        target_pixel_loss = torch.reshape(x, [-1])
```
Comment (puyuan1996) [inline ding/reward_model/count_based_model.py:346]: why named as target_pixel_loss? I guess x  should be the target pixel.

File: ding/reward_model/count_based_model.py [hunk 10]
```diff
@@ -0,0 +1,354 @@
...
+        Args:
+            activation_fn: Activation to use for the top 1/2 input channels.
+        """
+        super().__init__()
+        self._activation_fn = activation_fn
+
+    def forward(self, x):
+        _, c, _, _ = x.shape
+        assert c % 2 == 0, "x must have an even number of channels."
+        x, gate = x[:, : c // 2, :, :], x[:, c // 2 :, :, :]
```
Comment (puyuan1996) [inline ding/reward_model/count_based_model.py:168]: wolud you please add some annataions and reference about the GatedActivation and GatedPixelCNNLayer? Thanks a lot!

File: ding/reward_model/count_based_model.py (hunk 8 above)
Comment (youngzhou1999) [reply ding/reward_model/count_based_model.py:108]: That's easy to see for case t = 0, where the pred_gain has no use if using t not t+1. And I'll add some annotations, thanks.

File: ding/reward_model/count_based_model.py (hunk 9 above)
Comment (youngzhou1999) [reply ding/reward_model/count_based_model.py:346]: You are right! In the `_probs` function I returned it as `target` as well. I add `_loss` here just to remind me it's used in loss function.

File: ding/reward_model/count_based_model.py (hunk 8 above)
Comment (puyuan1996) [reply ding/reward_model/count_based_model.py]: Ok, got it. Thanks.

File: ding/reward_model/count_based_model.py [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+        self._mask_center = mask_center
+
+        # Vertical stack convolutions.
+        self._vstack_1xN = nn.Conv2d(
+            in_channels=self._in_channels,
+            out_channels=self._out_channels,
+            kernel_size=(1, self._kernel_size),
+            padding=(0, self._padding),
+        )
+        # TODO(eugenhotaj): Is it better to shift down the the vstack_Nx1 output
```
Comment (PaParaZz1) [inline ding/reward_model/count_based_model.py:209]: remove other code author information

File: ding/reward_model/count_based_model.py [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+            padding=(0, self._padding + int(self._mask_center)),
+        )
+        self._hstack_residual = nn.Conv2d(
+            in_channels=out_channels, out_channels=out_channels, kernel_size=1
+        )
+        self._hstack_skip = nn.Conv2d(
+            in_channels=out_channels, out_channels=out_channels, kernel_size=1
+        )
+
+    def forward(self, vstack_input, hstack_input):
```
Comment (PaParaZz1) [inline ding/reward_model/count_based_model.py:241]: Please obey the same comment format and add shapes information for neural network implementation, like [this](https://github.com/opendilab/DI-engine/blob/main/ding/model/template/q_learning.py#L86)

File: ding/reward_model/count_based_model.py [hunk 13]
```diff
@@ -0,0 +1,433 @@
...
+        '''
+        Overview:
+            Using input data to train the network while estimating intrintic reward.
+        Arguments:
+            train_data (:obj:'torch.Tensor'): Observation with shape [1, 1, 42, 42].
+        '''
+        flattened_logits, target_pixel_loss, _ = self._counter(train_data)
+
+        # flattened_logits shape: [BHWC, D]; target_pixel_loss shape: [BHWC]
+        loss = nn.CrossEntropyLoss(reduction='none')(
```
Comment (PaParaZz1) [inline ding/reward_model/count_based_model.py:96]: You should check the image input range of this reward model is in [0, 1], add its target should multiply 7(with pixel values quantized to 8 bins in original paper) and then transform into long type

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]
